### PR TITLE
Transition delivery payment

### PR DESCRIPTION
**Pull request by kanishkablack**
## Motivation and Context

A transition of order status from `delivery` to `payment`.

## Describe your changes

- Capture payment info from the context
- Computer packages total
- Creates a payment
- CheckOut Payment method
- Related test cases added

## Any further comments?

https://www.pivotaltracker.com/story/show/158682466

------

## Checklist

- [x] I have read [CONTRIBUTING.md][contributing].
- [x] My code follows the [style guidelines][contributing] of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [ ] I have updated the documentation wherever necessary.
- [ ] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

------
Dear Gatekeeper,

Please make sure that the commits that will be merged or rebased into the base
branch are [formatted according to our standards][commit-style].

> The only additional requirement is that the the PR number must appear in the
> commit title in brackets: `(#XXX)`

[commit-style]: https://github.com/aviabird/snitch/blob/develop/CONTRIBUTING.md#styling

### CODE REVIEW

File: apps/snitch_core/lib/core/data/model/payment/payment.ex [hunk 1]
```diff
@@ -54,4 +55,18 @@ defmodule Snitch.Data.Model.Payment do
   def to_subtype(%Payment{payment_type: "ccd"} = payment) do
     CardPaymentModel.from_payment(payment.id)
   end
+
+  def compute_order_payments(order) do
```
Comment (oyeb) [inline apps/snitch_core/lib/core/data/model/payment/payment.ex]: So this function reads like this: `Model.Payment.compute_order_payments/1`.
How about: `Model.Payment.total/1`?

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 2]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
...
+
+    ### 3rd party
+
+    * Payubiz
+
+    
+
+  """
+  @spec compute_order_payment(Context.t()) :: Context.t()
+  def compute_order_payment(
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: `Transitions.compute_order_payment` vs `Transitions.save_payment` / `Transitions.save_payment_info`

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 3]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
         {:halt, error}
     end)
   end
+
+  # Guard Class to check negative amount is skipped.
+
+  defp calc_payable_amount(order) do
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: This should be in `Model.Order`, and name this better, simple `payable_amount/1` works best.

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 4]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
...
+    total_amount = Money.add!(packages_total, order.total)
+    amount_paid_before = Payment.compute_order_payments(order)
+    Money.sub!(total_amount, amount_paid_before)
+  end
+
+  defp process_payment_chk(context, multi, payment_method, payment, order, amount_to_pay) do
+    params = %{
+      amount: amount_to_pay,
+      order_id: order.id,
+      slug: "check-payment"
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: Umm... there's a unique_constraint on `slug`. Could you change the `Payment` schema to use the new `Ecto.Nanoid`, also rename the `:slug` column to `:number`.

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 5]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
...
+    function = fn _ ->
+      CheckPaymentModel.create(params)
+    end
+
+    struct(context, multi: Multi.run(multi, :checkpayment, function))
+  end
+
+  defp process_payment_ccd(context, multi, payment_method, payment, order, amount_to_pay) do
+    case validate_payment_method(payment) do
+      %{changes: %{card: %{errors: []}}} ->
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: You can simply match on the `valid?` key.

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 6]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
...
+
+  @doc """
+  Validator for payment method
+
+  Responses
+  {:ok, payment}
+  {:error, errors}
+  """
+
+  defp validate_payment_method(payment) do
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: Why is this necessary? If the changeset is invalid, the multi will fail when it runs anyways, pls remove this, and the related complexity in `process_payment_ccd`

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 7]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
...
+          month: 12,
+          name_on_card: "Helper",
+          brand: "VISA",
+          number: "411141114114111",
+          card_name: "Rupay",
+          user_id: "Somer user id"
+        }
+    }
+
+    `user_id` needs to be injected into map through webcontoller.
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: Why? AFAIK no payment schema needs this.

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 8]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
...
+    
+
+  """
+  @spec compute_order_payment(Context.t()) :: Context.t()
+  def compute_order_payment(
+        %Context{valid?: true, struct: %Order{id: order_id} = order} = context
+      ) do
+    %{state: %{payment: payment}, multi: multi} = context
+
+    amount_to_pay = calc_payable_amount(order)
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: This is totally wrong. Payment records have `:state`.
```elixir
payable_amount = Repo.one(from p in Payment, where: p.order_id == ^order.id and p.state == "paid", select: sum(p.total))
```

File: apps/snitch_core/lib/core/domain/order/transitions.ex [hunk 9]
```diff
@@ -194,4 +196,118 @@ defmodule Snitch.Domain.Order.Transitions do
...
+    payment_method = PaymentMethod.get(payment.payment_method_id)
+
+    case payment_method.code do
```
Comment (oyeb) [inline apps/snitch_core/lib/core/domain/order/transitions.ex]: It is much cleaner to make a function with 2 clauses to handle this, since:
1. you've already made 2 functions
2. the `case` can be removed

File: apps/snitch_core/test/domain/order/transitions_test.exs [hunk 10]
```diff
@@ -199,4 +219,125 @@ defmodule Snitch.Domain.Order.TransitionsTest do
...
+
+    setup %{embedded_shipping_methods: methods} do
+      order = insert(:order, user: build(:user), total: Money.new(3, :USD))
+
+      [order: order, packages: [insert(:package, shipping_methods: methods, order: order)]]
+    end
+
+    @tag shipping_method_count: 1
+    test "with valid card details", %{order: order, packages: [package], shipping_methods: [sm]} do
+      expect(Snitch.Tools.DefaultsMock, :fetch, fn :currency -> {:ok, :USD} end)
```
Comment (oyeb) [inline apps/snitch_core/test/domain/order/transitions_test.exs]: 1. Use the `expect/4` clause, save a few lines.
2. Why are 2 transitions being tested here?

File: apps/snitch_core/test/domain/order/transitions_test.exs [hunk 11]
```diff
@@ -199,4 +219,125 @@ defmodule Snitch.Domain.Order.TransitionsTest do
...
+        |> Context.new(state: %{payment: payment})
+        |> Transitions.compute_order_payment()
+
+      assert result.valid?
+      assert {:ok, %{cardpayment: %{payment: payment}}} = Repo.transaction(result.multi)
+      assert Money.reduce(payment.amount) == Money.add!(order.total, package.total)
+    end
+
+    @tag shipping_method_count: 1
+    test "with out valid card details", %{
```
Comment (oyeb) [inline apps/snitch_core/test/domain/order/transitions_test.exs]: "with invalid card details"

**Comment by oyeb** [comment]
@ashish173: The PayUBiz schema is blocking this feature in a big way. Any progress on that?

File: apps/snitch_core/lib/core/domain/order/transitions.ex (hunk 6 above)
Comment (kanishkablack) [reply apps/snitch_core/lib/core/domain/order/transitions.ex]: Just need minor clarification.
If a user `card` payment method id and  enter card details
does the card details enter transition ? or will be controller handles the card details ?

My assumption, that card details are handled in transition.

If card details are handled by controller, we can remove couple of functions.
